### PR TITLE
[Feature] Introduces `self.signer`.

### PR DESCRIPTION
**Pull request by d0cd**
This PR introduces `self.signer` as a valid operand in a function.
`self.caller` refers to the immediate caller of the function, while the existing `self.signer` refers to the user that originated the sequence of calls leading up to the currently executing function.

Some interesting points:
- By asserting `assert.eq self.caller self.signer;`, a developer can restrict their function such that it can only be called by users.
- By asserting `assert.neq self.caller self.signer;` a developer can restrict their function such that it can only be called by other programs.

Related Work:
- https://github.com/AleoHQ/ARCs/tree/master/arc-0030
- #1178 
- #1969 
- #2030

### CODE REVIEW

File: synthesizer/process/src/stack/register_types/matches.rs [hunk 1]
```diff
@@ -74,26 +74,17 @@ impl<N: Network> RegisterTypes<N> {
...
-                    let caller_type = PlaintextType::Literal(LiteralType::Address);
-                    // Ensure the caller type matches the member type.
-                    ensure!(
-                        &caller_type == member_type,
-                        "Struct member '{struct_name}.{member_name}' expects {member_type}, but found '{caller_type}' in the operand '{operand}'.",
+                        &operand_type == member_type,
+                        "Struct member '{struct_name}.{member_name}' expects {member_type}, but found '{operand_type}' in the operand '{operand}'.",
                     )
                 }
+                // Ensure the parent type (address
```
Comment (ljedrz) [inline synthesizer/process/src/stack/register_types/matches.rs]: something seems to have been cut here

File: circuit/program/src/request/verify.rs [hunk 2]
```diff
@@ -19,17 +19,19 @@ impl<A: Aleo> Request<A> {
...
+    ///     challenge' := HashToScalar(r * G, pk_sig, pr_sig, caller, \[tvk, tcm, parent, is_root, function ID, input IDs\])
     pub fn verify(&self, input_types: &[console::ValueType<A::Network>], tpk: &Group<A>) -> Boolean<A> {
         // Compute the function ID as `Hash(network_id, program_id, function_name)`.
         let function_id = A::hash_bhp1024(
             &(&self.network_id, self.program_id.name(), self.program_id.network(), &self.function_name).to_bits_le(),
         );
 
-        // Construct the signature message as `[tvk, tcm, function ID, input IDs]`.
+        // Construct the signature message as `[tvk, tcm, parent, is_root, function ID, input IDs]`.
         let mut message = Vec::with_capacity(3 + 4 * self.input_ids.len());
```
Comment (howardwu) [inline circuit/program/src/request/verify.rs]:
```suggestion
        let mut message = Vec::with_capacity(5 + 4 * self.input_ids.len());
```

File: console/program/src/request/serialize.rs [hunk 3]
```diff
@@ -21,8 +21,10 @@ impl<N: Network> Serialize for Request<N> {
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         match serializer.is_human_readable() {
             true => {
-                let mut transition = serializer.serialize_struct("Request", 9)?;
+                let mut transition = serializer.serialize_struct("Request", 11)?;
```
Comment (howardwu) [inline console/program/src/request/serialize.rs]:
```suggestion
                let mut transition = serializer.serialize_struct("Request", 13)?;
```

File: console/program/src/request/sign.rs [hunk 4]
```diff
@@ -69,10 +71,13 @@ impl<N: Network> Request<N> {
             &(U16::<N>::new(N::ID), program_id.name(), program_id.network(), function_name).to_bits_le(),
         )?;
 
-        // Construct the hash input as `(r * G, pk_sig, pr_sig, caller, [tvk, tcm, function ID, input IDs])`.
+        // Map `is_root` to a field element.
+        let is_root_field = Field::ternary(&is_root, &Field::one(), &Field::zero());
+
+        // Construct the hash input as `(r * G, pk_sig, pr_sig, caller, [tvk, tcm, parent, is_root, function ID, input IDs])`.
         let mut message = Vec::with_capacity(5 + 2 * inputs.len());
```
Comment (howardwu) [inline console/program/src/request/sign.rs]:
```suggestion
        let mut message = Vec::with_capacity(9 + 2 * inputs.len());
```

File: console/program/src/request/verify.rs [hunk 5]
```diff
@@ -71,10 +80,15 @@ impl<N: Network> Request<N> {
             }
         };
 
-        // Construct the signature message as `[tvk, tcm, function ID, input IDs]`.
+        // Convert `self.is_root` to a field element.
+        let is_root = if *self.is_root { Field::one() } else { Field::zero() };
+
+        // Construct the signature message as `[tvk, tcm, parent, is_root, function ID, input IDs]`.
         let mut message = Vec::with_capacity(1 + self.input_ids.len());
```
Comment (howardwu) [inline console/program/src/request/verify.rs]:
```suggestion
        let mut message = Vec::with_capacity(5 + self.input_ids.len());
```

File: synthesizer/snark/src/universal_srs.rs [hunk 6]
```diff
@@ -35,6 +35,9 @@ impl<N: Network> UniversalSRS<N> {
         #[cfg(feature = "aleo-cli")]
         let timer = std::time::Instant::now();
 
+        println!("unversal_srs: {function_name}");
+        println!("   num_public: {}", assignment.num_public());
+
```
Comment (howardwu) [inline synthesizer/snark/src/universal_srs.rs]:
```suggestion
```

File: synthesizer/src/vm/mod.rs [hunk 7]
```diff
@@ -242,7 +242,10 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
         // Ensure the block is valid genesis block.
         match block.is_genesis() {
             true => Ok(block),
-            false => bail!("Failed to initialize a genesis block"),
+            false => {
+                println!("Foo");
+                bail!("Failed to initialize a genesis block")
+            }
```
Comment (howardwu) [inline synthesizer/src/vm/mod.rs]:
```suggestion
            false => bail!("Failed to initialize a genesis block"),
```

**Review by howardwu — APPROVED** [review]

**Pull request merged**